### PR TITLE
A couple of improvements

### PR DESCRIPTION
**Pull request by bjorn3**
See the individual commits for why each change is an improvement in my opinion.

### CODE REVIEW

**Comment by l0kod** [comment]
Thanks for these improvements!

My only concern is about the `AsRawFd` replacement with `AsFd` and `OwnedFd` which are not available with current Rust editions, but only starting with Rust 1.63. Could we conditionally use this type when it is available with supported Rust versions?

**Comment by bjorn3** [comment]
I will drop the `AsFd` change for now as switching between `AsRawFd` and `AsFd` is a breaking change, so can't be done depending on the rust version. For future reference this commit was 4a3226322b09135b19354ff776e95b13ecca4ec6.

**Comment by l0kod** [comment]
> I will drop the `AsFd` change for now as switching between `AsRawFd` and `AsFd` is a breaking change, so can't be done depending on the rust version. For future reference this commit was [4a32263](https://github.com/landlock-lsm/rust-landlock/commit/4a3226322b09135b19354ff776e95b13ecca4ec6).

I'm improving this crate for a new major release, so I don't mind breaking the API now (especially this small), but it seems reasonable to be compatible with at least the latest Rust edition. I tested this commit and it passes all the tests without changing the Rust edition, so I guess it's not an issue (as long as rustc >= 1.63, which should be fine) and it will still be compatible with the Rust 2018 and 2021 editions, right?

**Comment by l0kod** [comment]
Anyway, the `cargo doc --workspace --all-features` (not part of the CI yet) threw a warning with the `AsFd` commit.

**Comment by bjorn3** [comment]
Nothing in this PR should cause any MSRV increase.

> I'm improving this crate for a new major release, so I don't mind breaking the API now (especially this small)

I mean that if I use an AsFd bound when rustc >=1.63 is detected and AsRawFd otherwise, then the API will break between those cases, so you can have code which works on rustc 1.62, but breaks on rustc 1.63 due to rust-landlock automatically switching from AsRawFd to AsFd.

> I tested this commit and it passes all the tests without changing the Rust edition, so I guess it's not an issue (as long as rustc >= 1.63, which should be fine) and it will still be compatible with the Rust 2018 and 2021 editions, right?

In rust editions allow breaking syntactic changes to be made in a way that allows older code to keep working on newer rustc versions by staying on an older edition. Anything that can be enabled on older editions will be enabled on them. So if a new api in libstd is introduced in a future rustc version, all editions will be allowed to use it when compiling with this future rustc version. As such there is no point in staying on a rust edition older than the latest edition supported by the oldest rustc version you want to use. For example if you want to support rustc 1.56 you will have to stay on the 2018 edition, but if you only support rustc 1.57 (which introduced the 2021 edition) then you can use the 2021 edition without having any effect on users of your crate (as editions are invisible outside of the crate itself) or users who want to compile a project that uses your crate (as you already only support rustc 1.57 in this example)

**Comment by bjorn3** [comment]
By the way what is actually the oldest rustc version you want to keep supporting? If you haven't decided yet, here are some suggestions:

* 1.48 (rustc version shipped with debian 11, no landlock support)
* 1.54 (newest version of rustc that the mrustc alternative rust compiler can directly build)
* 1.61 (rustc version shipped with Ubuntu 22.04 LTS, has landlock v1 abi support)
* 1.63 (rustc version currently shipped with the upcoming debian 12, has landlock v2 abi support)
* 1.65 (previous rustc version)

**Comment by l0kod** [comment]
Thanks for these explanations.

OK, so the Rust edition doesn't imply to stick with a specific Rust version except for related syntax breaking changes, which is not the case to support `AsFd`. The remaining question is to pick a minimal Rust version to build this crate. Because `AsFd` is an important safety feature (that wasn't available when I started this crate), I'd very much like to choose the 1.63 Rust version as the minimal one. Furthermore at least Debian Sid and Arch Linux have Rust >= 1.63, and it is really easy to pick a newer version with rustup.

Please create a new PR with the `AsFd` commit. You can also update the readme to specify the MSRV.

**Comment by bjorn3** [comment]
Opened https://github.com/landlock-lsm/rust-landlock/pull/25.

File: src/ruleset.rs [hunk 1]
```diff
@@ -456,7 +456,6 @@ pub trait RulesetCreatedAttr: Sized + AsMut<RulesetCreated> {
         I: IntoIterator<Item = Result<T, E>>,
         T: Rule<U>,
         U: Access,
-        E: std::error::Error,
```
[the remark this answers is not in the document]
Comment (bjorn3) [reply src/ruleset.rs:459]: `anyhow::Error` dereferences to `dyn std::error::Error`, but it doesn't directly implement it. Turns out this implementation can't be added due to a conflict with two existing trait implementations: https://github.com/dtolnay/anyhow/issues/136#issuecomment-758897238

**Pull request merged**